embedded-hal-async/src/serial.rs

```diff
@@ -0,0 +1,74 @@
+//! Serial interface
+
+use core::future::Future;
+pub use embedded_hal::serial::{Error, ErrorKind};
+
+/// Read half of a serial interface
+///
+/// Some serial interfaces support different data sizes (8 bits, 9 bits, etc.);
+/// This can be encoded in this trait via the `Word` type parameter.
+pub trait Read<Word: 'static = u8> {
+    /// Read error
+    type Error: Error;
+
+    /// The future associated with the `read` method.
+    type ReadFuture<'a>: Future<Output = Result<(), Self::Error>> + 'a
+    where
+        Self: 'a;
+
+    /// Reads words from the serial interface into the supplied slice.
+    fn read<'a>(&'a mut self, read: &'a mut [Word]) -> Self::ReadFuture<'a>;
```

Comment:
i don't really want to derail this PR further / we can totally discuss elsewhere, but, it seems to me this needs to be able to return a _partially filled_ result.

say you're reading serial payloads of size up to N, to use this trait you're going to need to always poll with `&[Word; 1]` to guarantee you'll return, undermining the benefits of buffering / DMA.

i think in this case the trait would be more useful with `Result<usize, Self::Error>` that tells you how full the buffer is, then the HAL can choose to return on `EOF` / timeout / buffer full as is relevant for the hardware.

Reply:
The problem with that is it's not specified when it returns, so different impls will do different things, making it useless for drivers.

In Embassy we specified it [like this](https://github.com/embassy-rs/embassy/blob/master/embassy-traits/src/uart.rs): 
- `read` returns on buffer full.
- `read_until_idle` returns on buffer full, or when at least 1 byte has been received and then the line goes idle.

`read_until_idle` is a separate trait because some chips (nrf) can't do it natively so they need a hardware timer to do it. Forcing everyone to waste a timer when not needed would be rude, so there's a [special impl](https://github.com/embassy-rs/embassy/blob/master/embassy-nrf/src/uarte.rs#L426) supporting it that you can opt-in to use.

in my experience, `read` is still useful if you know your packets have a fixed size, and it maps to a single DMA read operation which is the lowest overhead possible. 

(this also applies to the uart blocking traits, IMO we should have a way to do reads without `nb`. The current `nb` read trait works really badly in the nRF because it can *only* do DMA reads.)

Reply:
> The problem with that is it's not specified when it returns, so different impls will do different things, making it useless for drivers.

i don't think adding `this _may_ return a partial buffer on timeout or idle` is useless? this is how reads under linux typically work, it just implies the need for a check and buffering at the driver level if you're not doing byte-wise processing.

> ... some chips (nrf) can't do it natively so they need a hardware timer to do it

rude of them (nrf) imo

> in my experience, read is still useful if you know your packets have a fixed size, and it maps to a single DMA read operation which is the lowest overhead possible.

the problem with this approach appears when you know your packet sizes but you somehow end up out-of-sync or with corrupted data, then you have to have application timeouts to recover anyway.